src/cls/rbd/cls_rbd.cc

```diff
@@ -229,6 +242,233 @@ static bool is_valid_id(const string &id) {
+
+  if (r < 0) {
+    return r;
+  }
+
+  return 0;
+}
+
+int image_remove_cg_ref(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "image_remove_cg_ref");
+  std::string cg_id;
+  int64_t pool_id;
+  try {
+    bufferlist::iterator iter = in->begin();
+    ::decode(cg_id, iter);
+    ::decode(pool_id, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  map<string, bufferlist> existing_refs;
+
+  int r = cls_cxx_map_get_vals(hctx, "", RBD_CG_REF_KEY, RBD_MAX_KEYS_READ, &existing_refs);
+  if (r < 1) {
+    return -ENOENT;
+  }
+
+  bufferlist refbl;
+  r = cls_cxx_map_get_val(hctx, RBD_CG_REF_KEY, &refbl);
+  if (r < 0) {
+    return r;
+  }
+
+  std::string ref_cg_id;
+  int64_t ref_pool_id;
+  bufferlist::iterator iter = refbl.begin();
+  ::decode(ref_pool_id, iter);
+  ::decode(ref_cg_id, iter);
+
+  if (ref_pool_id != pool_id || ref_cg_id != cg_id) {
+    return -EBADF;
+  }
+
+  r = cls_cxx_map_remove_key(hctx, RBD_CG_REF_KEY);
+  if (r < 0) {
+    return r;
+  }
+
+  return 0;
+}
+
+int cg_add_image(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "cg_add_image");
+
+  std::string image_id;
+  int64_t pool_id;
+  try {
+    bufferlist::iterator iter = in->begin();
+    ::decode(image_id, iter);
+    ::decode(pool_id, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  map<string, bufferlist> existing_refs;
+
+  string image_key = RBD_IMAGE_KEY_PREFIX + image_id + "_" + boost::lexical_cast<std::string>(pool_id);
+
+  bufferlist image_val_bl;
+  int64_t link_state = LINK_DIRTY;
+  ::encode(link_state, image_val_bl);
+  int r = cls_cxx_map_set_val(hctx, image_key, &image_val_bl);
+  if (r < 0) {
+    return r;
+  }
+
+  return 0;
+}
+
+int cg_dirty_link(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "cg_dirty_link");
+
+  std::string image_id;
+  int64_t pool_id;
+  try {
+    bufferlist::iterator iter = in->begin();
+    ::decode(image_id, iter);
+    ::decode(pool_id, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  string image_key = RBD_IMAGE_KEY_PREFIX + image_id + "_" + boost::lexical_cast<std::string>(pool_id);
+
+  bufferlist statebl;
+  int64_t new_state = LINK_DIRTY;
+  ::encode(new_state, statebl);
+  int r = cls_cxx_map_set_val(hctx, image_key, &statebl);
+  if (r < 0) {
+    return r;
+  }
+
+  return 0;
+}
+
+int cg_to_default(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "cg_to_default");
+
+  std::string image_id;
+  int64_t pool_id;
+  try {
+    bufferlist::iterator iter = in->begin();
+    ::decode(image_id, iter);
+    ::decode(pool_id, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+  string image_key = RBD_IMAGE_KEY_PREFIX + image_id + "_" + boost::lexical_cast<std::string>(pool_id);
+
+  int64_t link_state;
+
+  int r = read_key(hctx, image_key, &link_state);
```

Comment:
You read the link_state but you don't do anything with it. Maybe you can avoid reading it.